### PR TITLE
Add convenience constructors for SliverList

### PR DESCRIPTION
**Pull request by thkim1011**
This PR adds a 
* `SliverList.builder` constructor that automatically constructs a `SliverChildBuilderDelegate` from an `itemBuilder` and passes it into the super constructor.
* `SliverList.separated` constructor (SliverList version of ListView.separated).
* `SliverFixedExtentList.builder` constructor that takes an extent as well.
* `SliverPrototypeExtentList.builder` that determines extent via a prototype element.

Related #113107.

### CODE REVIEW

File: packages/flutter/lib/src/widgets/sliver.dart [hunk 1]
```diff
@@ -1035,6 +1035,24 @@ class SliverList extends SliverMultiBoxAdaptorWidget {
     required super.delegate,
   });
 
+  /// Creates a sliver list from an item builder.
```
Comment (Piinks) [inline packages/flutter/lib/src/widgets/sliver.dart:1038]: The ListView.builder docs could probably be re-used here. There are examples of re-using docs like this if you search for `{@template...` as well as `{@macro...` in the framework.

Comment (Piinks) [reply packages/flutter/lib/src/widgets/sliver.dart:1038]: It might be nice to include a sample snippet of code here that shows it in use.

**Review by Piinks — COMMENTED** [review]
We chatted about adding 

- SliverList.separated like ListView.separated, 

but also if you look at ListView.buildChildLayout, you'll see

- SliverFixedExtentList
- and SliverPrototypeExtentList. 

How do you feel about including those here as well? That way, all the ListView slivers are accounted for. :)

File: packages/flutter/lib/src/widgets/sliver.dart (hunk 1 above)
Comment (thkim1011) [reply packages/flutter/lib/src/widgets/sliver.dart]: Done!

File: packages/flutter/lib/src/widgets/sliver.dart [hunk 2]
```diff
@@ -1035,6 +1036,146 @@ class SliverList extends SliverMultiBoxAdaptorWidget {
...
+  ///
+  /// {@macro flutter.widgets.PageView.findChildIndexCallback}
+  ///
+  /// The `addAutomaticKeepAlives` argument corresponds to the
+  /// [SliverChildBuilderDelegate.addAutomaticKeepAlives] property. The
+  /// `addRepaintBoundaries` argument corresponds to the
+  /// [SliverChildBuilderDelegate.addRepaintBoundaries] property. The
+  /// `addSemanticIndexes` argument corresponds to the
+  /// [SliverChildBuilderDelegate.addSemanticIndexes] property.
+  /// /// {@tool snippet}
```
Comment (Piinks) [inline packages/flutter/lib/src/widgets/sliver.dart:1062]: Nits:
```suggestion
  ///
  /// {@tool snippet}
```

File: packages/flutter/lib/src/widgets/sliver.dart [hunk 3]
```diff
@@ -1035,6 +1036,146 @@ class SliverList extends SliverMultiBoxAdaptorWidget {
...
+  ///
```
Comment (Piinks) [inline packages/flutter/lib/src/widgets/sliver.dart:1063]:
```suggestion
```

File: packages/flutter/lib/src/widgets/sliver.dart [hunk 4]
```diff
@@ -1035,6 +1036,146 @@ class SliverList extends SliverMultiBoxAdaptorWidget {
...
+  /// The `addAutomaticKeepAlives` argument corresponds to the
+  /// [SliverChildBuilderDelegate.addAutomaticKeepAlives] property. The
+  /// `addRepaintBoundaries` argument corresponds to the
+  /// [SliverChildBuilderDelegate.addRepaintBoundaries] property. The
+  /// `addSemanticIndexes` argument corresponds to the
+  /// [SliverChildBuilderDelegate.addSemanticIndexes] property.
+  /// {@tool snippet}
+  /// 
+  //  This example shows how to create [SliverList] whose [Container] items
```
Comment (Piinks) [inline packages/flutter/lib/src/widgets/sliver.dart:1125]:
```suggestion
  //  This example shows how to create a [SliverList] whose [Container] items
```

File: packages/flutter/lib/src/widgets/sliver.dart [hunk 5]
```diff
@@ -1035,6 +1036,146 @@ class SliverList extends SliverMultiBoxAdaptorWidget {
...
+  ///       color: Colors.lightBlue[100 * (index % 9)],
+  ///       child: Text('list item $index'),
+  ///     );
+  ///   },
+  ///   separatorBuilder: (BuildContext context, int index) => const Divider(),
+  /// )
+  /// ```
+  /// {@end-tool}
+  SliverList.separated({
+     super.key,
```
Comment (Piinks) [inline packages/flutter/lib/src/widgets/sliver.dart:1142]:
```suggestion
    super.key,
```

File: packages/flutter/lib/src/widgets/sliver_prototype_extent_list.dart [hunk 6]
```diff
@@ -40,6 +40,71 @@ class SliverPrototypeExtentList extends SliverMultiBoxAdaptorWidget {
...
+  ///
+  /// {@macro flutter.widgets.PageView.findChildIndexCallback}
+  ///
+  /// The `addAutomaticKeepAlives` argument corresponds to the
+  /// [SliverChildBuilderDelegate.addAutomaticKeepAlives] property. The
+  /// `addRepaintBoundaries` argument corresponds to the
+  /// [SliverChildBuilderDelegate.addRepaintBoundaries] property. The
+  /// `addSemanticIndexes` argument corresponds to the
+  /// [SliverChildBuilderDelegate.addSemanticIndexes] property.
+  /// {@tool snippet}
```
Comment (Piinks) [inline packages/flutter/lib/src/widgets/sliver_prototype_extent_list.dart:67]:
```suggestion
  ///
  /// {@tool snippet}
```

File: packages/flutter/lib/src/widgets/sliver_prototype_extent_list.dart [hunk 7]
```diff
@@ -40,6 +40,71 @@ class SliverPrototypeExtentList extends SliverMultiBoxAdaptorWidget {
...
+  }) : super(
+         delegate: SliverChildBuilderDelegate(
+           itemBuilder,
+           findChildIndexCallback: findChildIndexCallback,
+           childCount: itemCount,
+           addAutomaticKeepAlives: addAutomaticKeepAlives,
+           addRepaintBoundaries: addRepaintBoundaries,
+           addSemanticIndexes: addSemanticIndexes,
+         ),
+       );
```
Comment (Piinks) [inline packages/flutter/lib/src/widgets/sliver_prototype_extent_list.dart:106]: Nit: just to be consistent with the rest
```suggestion
  }) : super(delegate: SliverChildBuilderDelegate(
         itemBuilder,
         findChildIndexCallback: findChildIndexCallback,
         childCount: itemCount,
         addAutomaticKeepAlives: addAutomaticKeepAlives,
         addRepaintBoundaries: addRepaintBoundaries,
         addSemanticIndexes: addSemanticIndexes,
       ));
```

**Review by Piinks — COMMENTED** [review]
Oh I just remembered! Can you add
- SiverList.list
- SliverFixedExtentList.list
- SliverPrototypeExtentList.list

These would cover the ListView default, where it creates a SliverChildListDelegate instead of a SliverChildBuilderDelegate

https://github.com/flutter/flutter/blob/60635be592eb9b24f787592255a4a213a803406e/packages/flutter/lib/src/widgets/scroll_view.dart#L1093-L1098

Maybe `.list` is not the best name, so if you think of another name, feel free to try it out! :)

File: packages/flutter/lib/src/widgets/sliver.dart (hunk 2 above)
Comment (thkim1011) [reply packages/flutter/lib/src/widgets/sliver.dart]: Fixed!

File: packages/flutter/lib/src/widgets/sliver.dart (hunk 3 above)
Comment (thkim1011) [reply packages/flutter/lib/src/widgets/sliver.dart:1062]: Fixed!

File: packages/flutter/lib/src/widgets/sliver.dart (hunk 4 above)
Comment (thkim1011) [reply packages/flutter/lib/src/widgets/sliver.dart]: Fixed!

File: packages/flutter/lib/src/widgets/sliver.dart (hunk 5 above)
Comment (thkim1011) [reply packages/flutter/lib/src/widgets/sliver.dart]: Fixed!

File: packages/flutter/lib/src/widgets/sliver_prototype_extent_list.dart (hunk 6 above)
Comment (thkim1011) [reply packages/flutter/lib/src/widgets/sliver_prototype_extent_list.dart]: Fixed!

**Review by Piinks — APPROVED** [review]
Flutter_LGTM

**Review by Piinks — COMMENTED** [review]
It looks like there are some trailing spaces causing the analyzer to fail, can you take a look?

**Pull request merged**